### ISSUE TITLE
Arrow-native backends

### ISSUE THREAD

**Issue by kszucs** [ibis-project/ibis#2798]
Since the first implementation of Ibis's Parquet backend the Arrow ecosystem has evolved a lot allowing tighter and more efficient integration between the two projects. A non-exhaustive list of projects and new components:
- DuckDB
- BlazingSQL
- ClickHouse support for Arrow-native formats
- Arrow C++ Datasets
- Arrow DataFusion and Ballista
- Arrow Flight
- Work in progress Arrow C++ Query Engine

**Comment by kszucs** [comment]
@datapythonista copied from https://github.com/ibis-project/ibis/issues/2795#issuecomment-851100974

>  SQL baseds should be implemented using SQLAlchemy. Feels like the Ibis engine used by Impala, Clickhouse, Bigquery is just reinventing the wheel. 

Indeed, SQLAlchemy's translation layer is probably more sophisticated. Actually both [clickhouse](https://github.com/xzkostyan/clickhouse-sqlalchemy) and [bigquery](https://github.com/googleapis/python-bigquery-sqlalchemy) have a sqlalchemy dialect.

> I think there are historical reasons for the way things are, but for new backends I think it makes more sense to implement a SQLAlchemy engine first, and then should be trivial to have it working in Ibis (particularly after the refactoring).

Can we support multiple output formats from sqlalchemy dialects? On long term we would like to retrieve the resultset directly in arrow-native formats without converting it to pandas. 

> I think at the moment is better to wait few more weeks. It's being tricky to get tests passing for the refactoring, but once those PRs are merged, I'll be able to follow up with easier ones that will delete a lot of repetitive code. Also, I'll start creating an ABC for a Connection class that will replace the Client and Backend ones. When that is done, implementing a backend should be way easier, since you'll have the spec of what's required (and eventually what's optional).

Does your refactoring equally affect both the SQL and File backends?

**Comment by datapythonista** [comment]
Sounds great. We've got #2627 for DataFusion.

OmniSci was also interested on being able to fetch data into Arrow, without converting to pandas. I'm fine with both, supporting multiple formats, or just support Arrow. I think users can simply `expr.execute.to_pandas()`, so not sure if it's worth supporting multiple in-memory formats.

Afaik file backends aren't really backends, but I/O and syntactic sugar on top of the pandas backend. Would be really nice to eventually have something like an bis expression with a filter backed by csv, and just load into memory the lines that match the filter. But that's not what's happening now. Data are loaded to pandas, and then the pandas backend is used. Answering your question, the parquet "backend" is 120 lines of code. Will be affected by the refactoring, since I'm merging `Backend` and `Client`, but nothing significant in practice.

**Comment by kszucs** [comment]
> I'm fine with both, supporting multiple formats, or just support Arrow. I think users can simply `expr.execute.to_pandas()`, so not sure if it's worth supporting multiple in-memory formats.

We should create another ticket for it with a more polite API change.

> Afaik file backends aren't really backends, but I/O and syntactic sugar on top of the pandas backend. Would be really nice to eventually have something like an bis expression with a filter backed by csv, and just load into memory the lines that match the filter. 

Yes, the long-term plan is to offload as much work as possible to the query engines and spare converting back-and-forth between pandas and arrow. 

> Will be affected by the refactoring, since I'm merging `Backend` and `Client`, but nothing significant in practice.

Would it make sense to prioritize non-sql backends until the `Backend` refactoring moves forward? In case of `DataFusion` we can translate `Ibis expressions` to `DataFusion expressions` instead of using the existing compilers.

**Comment by datapythonista** [comment]
> Would it make sense to prioritize non-sql backends until the `Backend` refactoring moves forward? In case of `DataFusion` we can translate `Ibis expressions` to `DataFusion expressions` instead of using the existing compilers.

Good point. Non-SQL backends shouldn't be as much affected., so sounds reasonable.

It's not clear to me why with datafusion is better to implement the backend with the Python API, than with the SQL API. You clearly know better than me, not questioning it, but would be nice if you can share a bit the reasons.

In my mind, implementing a SQL backend in Ibis should be trivial, it's mainly defining the map of operations. I think we could even make all the standard SQL work without a backend, and use a default map. While using a Python API is necessarily a lot of work. Unless we have a base implementation for libraries using the pandas API which would be useful for dask and possibly others, but I don't think datafusion has a pandas-like API (and we are far from having that).

**Comment by kszucs** [comment]
> It's not clear to me why with datafusion is better to implement the backend with the Python API, than with the SQL API. You clearly know better than me, not questioning it, but would be nice if you can share a bit the reasons.

We can directly construct a datafusion logical plan using the python API rather than constructing a SQL query then passing it to DataFusion to parse it. It should be much simpler than the pandas backend since the logical plan is lazily evaluated.
Although I'd need to experiment with both APIs to clearly see their advantages.

> In my mind, implementing a SQL backend in Ibis should be trivial, it's mainly defining the map of operations. 

Agree, assuming that the particular backend has a well covered SQLAlchemy plugin (which is the future plan if I understand correctly). We possibly need to implement these plugins ourselves for certain backends (like DuckDB) maintained in separate repositories which brings additional maintenance cost.

> I think we could even make all the standard SQL work without a backend, and use a default map. While using a Python API is necessarily a lot of work. Unless we have a base implementation for libraries using the pandas API which would be useful for dask and possibly others, but I don't think datafusion has a pandas-like API (and we are far from having that).

Eventually we would like to support other query engines which don't necessarily provide a SQL API, one example is [Polars](https://github.com/pola-rs/polars).

**Comment by cpcloud** [comment]
Closing this out for now, we've got the sweet new datafusion backend in 2.1.0 (thanks @kszucs!).

Let's open up new issues for specific backends.
